misc/efi-stub/container.c

```diff
@@ -0,0 +1,488 @@
+#endif
+
+	if (err != EFI_SUCCESS) {
+		Print(L"Failed to allocate memory for ACRN HV %r\n", err);
+		goto out;
+	}
+
+	for (i = 0; i < ehdr->e_phnum; i++) {
+		phdr = (Elf32_Phdr *)((UINT8 *)pbase + i * ehdr->e_phentsize);
+		if ((phdr->p_type != PT_LOAD) || (phdr->p_memsz == 0) || (phdr->p_offset == 0)) {
+			continue;
+		}
+
+		if (phdr->p_filesz > phdr->p_memsz) {
+			err = EFI_LOAD_ERROR;
+			goto out;
+		}
+
+		addr = (EFI_PHYSICAL_ADDRESS)(*hv_hpa + (phdr->p_paddr - CONFIG_HV_RAM_START));
+		memcpy((char *)addr, (const char *)(elf_image + phdr->p_offset), phdr->p_filesz);
+
+		if (phdr->p_memsz > phdr->p_filesz) {
+			addr = (EFI_PHYSICAL_ADDRESS)(*hv_hpa + (phdr->p_paddr - CONFIG_HV_RAM_START + phdr->p_filesz));
+			(void)memset((void *)addr, 0x0, (phdr->p_memsz - phdr->p_filesz));
+		}
+	}
+
+out:
+	return err;
+}
+
+/**
+ * @brief Load hypervisor into memory from a container blob
+ * 
+ * @param[in] hvld Loader handle
+ * 
+ * @return EFI_SUCCESS(0) on success, non-zero on error
+ */
+static EFI_STATUS container_load_boot_image(HV_LOADER hvld)
+{
+	EFI_STATUS err = EFI_SUCCESS;	
+	struct container *ctr = (struct container *)hvld;
+
+	LOADER_COMPRESSED_HEADER *lzh = NULL;
+
+	/* hv_cmdline.txt: to be copied into memory by the fill_bootcmd_tag operation later */
+	lzh = ctr->lzh_ptr[LZH_BOOT_CMD];
+	ctr->boot_cmdsize = lzh->Size + StrnLen(ctr->options, ctr->options_size);	
+
+	/* acrn.32.out */
+	lzh = ctr->lzh_ptr[LZH_BOOT_IMG];	
+	err = load_acrn_elf((const UINT8 *)lzh->Data, &ctr->hv_hpa);
+	if (err != EFI_SUCCESS) {
+		Print(L"Failed to load ACRN HV ELF Image%r\n", err);
+		goto out;
+	}
+out:
+	return err;	
+}
+
+/**
+ * @brief Load kernel modules and acpi tables into memory from a container blob
+ * 
+ * @param[in] hvld Loader handle
+ * 
+ * @return EFI_SUCCESS(0) on success, non-zero on error
+ */
+ static EFI_STATUS container_load_modules(HV_LOADER hvld)
+{
+	EFI_STATUS err = EFI_SUCCESS;
+	struct container *ctr = (struct container *)hvld;	
+
+	UINTN i;
+	
+	UINT8 * p = NULL;
+	LOADER_COMPRESSED_HEADER *lzh = NULL;
+	
+	/* scan module headers to calculate required memory size to store files */
+	for (i = LZH_MOD0_CMD; i < ctr->lzh_count - 1; i++) {
+		if ((i % 2) == 0) {	/* vm0_tag.txt, vm1_tag.txt, acpi_vm0.txt ... */
+			ctr->total_modcmdsize += ctr->lzh_ptr[i]->Size;
+		} else {	/* vm0_kernel, vm1_kernel, vm0_acpi.bin ... */
+			ctr->total_modsize += ALIGN_UP(ctr->lzh_ptr[i]->Size, EFI_PAGE_SIZE);
+		}
+	}
+	/* exclude hypervisor and SBL signature files. e.g.)
+	 *    lzh_count = 9 (hv_cmdline, acrn.32.out, vm0_tag, vm0_kernel, vm1_tag, vm1_kernel, vm0_acpi_tag, vm0_acpi, sig)
+	 *    mod_count = 3 (vm0_tag + vm0_kernel, vm1_tag + vm1_kernel, vm0_acpi_tag + vm0_acpi)
+	 */
+	ctr->mod_count = (ctr->lzh_count - 3) / 2;
+	
+	/* allocate single memory region to store all binary files to avoid mmap fragmentation */
+	err = emalloc_reserved_aligned(&(ctr->mod_hpa), ctr->total_modsize, 
```

Comment:
RFC: Is it redundant? Currently I allocated memory to copy Kernel & ACPI table files from container to. Since all files are embedded into the ACRN EFI image which is already loaded by BIOS as part of the EfiLoaderCode, technically it's should be possible to directly pass the file locations to HV w/o data copying. Do you see any concerns to do so? For example, it's not a good manner that HV access Bootloaders's any code section.